packages/next-intl/src/middleware/middleware.tsx

```diff
@@ -44,11 +44,14 @@ export default function createMiddleware<Locales extends AllLocales>(
   const configWithDefaults = receiveConfig(config);
 
   return function middleware(request: NextRequest) {
+    // Resolve potential foreign symbols (e.g. /ja/%E7%B4%84 → /ja/約))
+    const nextPathname = decodeURI(request.nextUrl.pathname);
```

Comment:
Decoding an incoming request seems easier than encoding every segment of specified pathnames, where we have symbols like `[]` that should not be encoded.